File: nil/cmd/nil/internal/wallet/new.go
package wallet

import (
	"fmt"

	"github.com/NilFoundation/nil/nil/cmd/nil/internal/common"
	"github.com/NilFoundation/nil/nil/cmd/nil/internal/config"
	"github.com/NilFoundation/nil/nil/common/check"
	"github.com/NilFoundation/nil/nil/internal/types"
	"github.com/NilFoundation/nil/nil/services/cliservice"
	"github.com/spf13/cobra"
)

var defaultNewWalletAmount = types.NewValueFromUint64(100_000_000)

func NewCommand(cfg *common.Config) *cobra.Command {
	serverCmd := &cobra.Command{
		Use:   "new",
		Short: "Create a new wallet with some initial balance on the cluster",
		RunE: func(cmd *cobra.Command, args []string) error {
			return runNew(cmd, args, cfg)
		},
		SilenceUsage: true,
	}

	setFlags(serverCmd)

	return serverCmd
}

func setFlags(cmd *cobra.Command) {
	params.salt = *types.NewUint256(0)
	cmd.Flags().Var(
		&params.salt,
		saltFlag,
		"The salt for the wallet address calculation")

	cmd.Flags().Var(
		types.NewShardId(&params.shardId, types.BaseShardId),
		shardIdFlag,
		"Specify the shard ID to interact with",
	)

	cmd.Flags().Var(
		&params.feeCredit,
		feeCreditFlag,
		"The fee credit for wallet creation. If set to 0, it will be estimated automatically",
	)

	params.newWalletAmount = defaultNewWalletAmount
	cmd.Flags().Var(
		&params.newWalletAmount,
		amountFlag,
		"The initial balance (capped at 10'000'000). The deployment fee will be subtracted from this balance",
	)
}

func runNew(_ *cobra.Command, _ []string, cfg *common.Config) error {
	amount := params.newWalletAmount
	if amount.Cmp(defaultNewWalletAmount) > 0 {
		logger.Warn().
			Msgf("The specified balance (%s) is greater than the limit (%s). Decrease it.", &params.newWalletAmount, defaultNewWalletAmount)
		amount = defaultNewWalletAmount
	}

	faucet, err := common.GetFaucetRpcClient()
	if err != nil {
		return err
	}
	srv := cliservice.NewService(common.GetRpcClient(), cfg.PrivateKey, faucet)
	check.PanicIfNotf(cfg.PrivateKey != nil, "A private key is not set in the config file")
	walletAddress, err := srv.CreateWallet(params.shardId, &params.salt, amount, params.feeCredit, &cfg.PrivateKey.PublicKey)
	if err != nil {
		return err
	}

	if err := common.PatchConfig(map[string]interface{}{
		common.AddressField: walletAddress.Hex(),
	}, false); err != nil {
		logger.Error().Err(err).Msg("failed to update the wallet address in the config file")
	}

	if !config.Quiet {
		fmt.Print("New wallet address: ")
	}
	fmt.Println(walletAddress.Hex())
	return nil
}

File: clijs/src/commands/wallet/new.ts
import {
  bytesToHex,
  Faucet,
  type Hex,
  LocalECDSAKeySigner,
  waitTillCompleted,
  WalletV1,
} from "@nilfoundation/niljs";
import { BaseCommand } from "../../base.js";
import { ConfigKeys } from "../../common/config.js";
import { Flags } from "@oclif/core";
import logger from "../../logger.js";

export const DefualtNewWalletAmount = 100_000_000n;

export default class WalletNew extends BaseCommand {
  static override description = "Create a new wallet";

  static override examples = ["<%= config.bin %> <%= command.id %>"];

  static flags = {
    salt: Flags.integer({
      char: "s",
      description: "The salt for the wallet address calculation",
      required: false,
      default: 0,
    }),
    shardId: Flags.integer({
      char: "i",
      description: "Specify the shard ID(>= 1) to interact with",
      required: false,
      default: 1,
      parse: async (input) => {
        const parsed = Number.parseInt(input, 10);
        if (Number.isNaN(parsed)) {
          throw new Error("Shard ID must be a number");
        }
        if (parsed < 1) {
          throw new Error("Shard ID must be greater than or equal to 1");
        }
        return parsed;
      },
    }),
    feeCredit: Flags.integer({
      char: "f",
      description:
        "The fee credit for wallet creation. If set to 0, it will be estimated automatically",
      required: false,
      default: 0,
    }),
    amount: Flags.integer({
      char: "a",
      description:
        "The initial balance (capped at 100'000'000). The deployment fee will be subtracted from this balance",
      required: false,
      default: Number(DefualtNewWalletAmount),
    }),
  };

  public async run(): Promise<Hex> {
    const { flags } = await this.parse(WalletNew);

    if (flags.amount > DefualtNewWalletAmount) {
      logger.warn(
        `The specified balance (${flags.amount}) is greater than the limit (${DefualtNewWalletAmount}). The default value is used.`,
      );
      flags.amount = Number(DefualtNewWalletAmount);
    }

    const privateKey = this.cfg?.[ConfigKeys.PrivateKey];
    if (!privateKey) {
      throw new Error(
        "Private key not found in config. Perhaps you need to run 'keygen new' first?",
      );
    }

    const signer = new LocalECDSAKeySigner({
      privateKey: privateKey as Hex,
    });

    const pubkey = signer.getPublicKey();
    const wallet = new WalletV1({
      pubkey: pubkey,
      salt: BigInt(flags.salt),
      shardId: flags.shardId,
      client:
        this.rpcClient ??
        (() => {
          throw new Error("RPC client is not initialized");
        })(),
      signer,
    });
    const address = wallet.address;

    const faucet = new Faucet(this.rpcClient);

    logger.debug(`Withdrawing ${flags.amount} to ${address}`);
    const faucetHash = await faucet.withdrawTo(address, BigInt(flags.amount));
    await waitTillCompleted(this.rpcClient, bytesToHex(faucetHash));

    if (this.quiet) {
      await wallet.selfDeploy(true);
      this.log(address);
    } else {
      this.log("Deploying wallet...");
      const tx = await wallet.selfDeploy(true);
      this.log(`Successfully deployed wallet with tx hash: ${bytesToHex(tx)}`);
      this.log(`Wallet address: ${address}`);
    }
    this.configManager?.updateConfig(ConfigKeys.NilSection, ConfigKeys.Address, address);
    return address;
  }
}
